src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -2200,22 +2201,30 @@ QueryPushdownSqlTaskList(Query *query, uint64 jobId,
 			maxShardOffset = -1;
 		}
 
+		/*
+		 * For left joins we don't care about the shards pruned for the right hand side.
+		 * If the right hand side would prune to a smaller set we should still send it to
+		 * all tables of the left hand side. However if the right hand side is bigger than
+		 * the left hand side we don't have to send the query to any shard that is not
+		 * matching anything on the left hand side.
+		 *
+		 * Instead we will simply skip any RelationRestriction if it is an OUTER join and
+		 * the table is part of the non-outer side of the join.
+		 */
+		if (IsNonOuterTableOfOuterJoin(relationRestriction))
```

Comment:
I'm kind of wondering: What about inner joins (in subqueries)?

If one side of the join is pruned away then the other side can be as well. Is there an easy way to achieve that here?

Reply:
Do you have a simple example of such a query? I have a hard time understanding from your description above

Reply:
any query with an inner join in a subquery

After your patch:
```sql
postgres=#  explain SELECT * FROM (SELECT random() FROM test a JOIN test b ON (a.x = b.x) WHERE a.x IN (1,1) ) a;
┌─────────────────────────────────────────────────────────────────────────────────────────┐
│                                       QUERY PLAN                                        │
├─────────────────────────────────────────────────────────────────────────────────────────┤
│ Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                     │
│   Task Count: 32                                                                        │
│   Tasks Shown: One of 32                                                                │
│   ->  Task                                                                              │
│         Node: host=localhost port=9700 dbname=postgres                                  │
│         ->  Hash Join  (cost=38.54..85.69 rows=260 width=8)                             │
│               Hash Cond: (b.x = a_1.x)                                                  │
│               ->  Seq Scan on test_102008 b  (cost=0.00..32.60 rows=2260 width=4)       │
│               ->  Hash  (cost=38.25..38.25 rows=23 width=4)                             │
│                     ->  Seq Scan on test_102008 a_1  (cost=0.00..38.25 rows=23 width=4) │
│                           Filter: (x = ANY ('{1,1}'::integer[]))                        │
└─────────────────────────────────────────────────────────────────────────────────────────┘
(11 rows)

Time: 31.608 ms
postgres=#  explain SELECT * FROM (SELECT random() FROM test a LEFT JOIN test b ON (a.x = b.x) WHERE a.x IN (1,1) ) a;
┌─────────────────────────────────────────────────────────────────────────────────────────┐
│                                       QUERY PLAN                                        │
├─────────────────────────────────────────────────────────────────────────────────────────┤
│ Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                     │
│   Task Count: 1                                                                         │
│   Tasks Shown: All                                                                      │
│   ->  Task                                                                              │
│         Node: host=localhost port=9701 dbname=postgres                                  │
│         ->  Hash Right Join  (cost=38.54..85.69 rows=260 width=8)                       │
│               Hash Cond: (b.x = a_1.x)                                                  │
│               ->  Seq Scan on test_102009 b  (cost=0.00..32.60 rows=2260 width=4)       │
│               ->  Hash  (cost=38.25..38.25 rows=23 width=4)                             │
│                     ->  Seq Scan on test_102009 a_1  (cost=0.00..38.25 rows=23 width=4) │
│                           Filter: (x = ANY ('{1,1}'::integer[]))                        │
└─────────────────────────────────────────────────────────────────────────────────────────┘
(11 rows)

```

Reply:
Even though there is a small difference in the query they follow slightly different code paths.

The non-subquery version goes through our own join order planner which apparently causes the correct shards to be pruned. The subquery version follows the codepath we are changing here.

Historically we take the union of all shard placements. This patch changes that specifically for outerjoins by not unioning with the inner part of an outer join. My feeling says that a union is the wrong approach here, and instead we should find the overlap of all joins except the inner part of the outerjoin. As said, this is a feeling, and I cannot proof it nor disprove it at the moment.

I tried to keep this change as simple as possible to solve the `left join` case, and with that all the outerjoins as postgres doesn't have big difference between left and right as they often rewrite one to the other.

Do you think it would make sense to address the change in union vs overlap here as well? I will create.a quick non-optimized prototype to check the impact on our test suite to be a bit more informed about the correctness of overlapping the shard placements.

Reply:
did a quick test, and blindly taking the intersect causes a lot of test failures. Some clearly show incorrect results as their are too many shards pruned away: https://github.com/citusdata/citus/compare/fix/left-join-pruning...fix/join-union-to-overlap?expand=1